hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/linux/resources/CGroupsResourceCalculator.java

```diff
@@ -0,0 +1,346 @@
+ * features.
+ */
+public class CGroupsResourceCalculator extends ResourceCalculatorProcessTree {
+  enum Result {
+    Continue,
+    Exit
+  }
+  private static final String PROCFS = "/proc";
+  static final String CGROUP = "cgroup";
+  static final String CPU_STAT = "cpuacct.stat";
+  static final String MEM_STAT = "memory.usage_in_bytes";
+  static final String MEMSW_STAT = "memory.memsw.usage_in_bytes";
+  private static final String USER = "user ";
+  private static final String SYSTEM = "system ";
+
+  private static final Pattern CGROUP_FILE_FORMAT = Pattern.compile(
+      "^(\\d+):([^:]+):/(.*)$");
+  private final String procfsDir;
+  private CGroupsHandler cGroupsHandler;
+
+  private String pid;
+  private File cpuStat;
+  private File memStat;
+  private File memswStat;
+
+  private final long jiffyLengthMs;
+  private BigInteger processTotalJiffies = BigInteger.ZERO;
+  private final CpuTimeTracker cpuTimeTracker;
+  private Clock clock;
+
+  private final static Object LOCK = new Object();
+  private static boolean firstError = true;
+
+  /**
+   * Create resource calculator for all Yarn containers.
+   * @throws YarnException Could not access cgroups
+   */
+  public CGroupsResourceCalculator() throws YarnException {
+    this(null, PROCFS, ResourceHandlerModule.getCGroupsHandler(),
+        SystemClock.getInstance());
+  }
+
+  /**
+   * Create resource calculator for the container that has the specified pid.
+   * @param pid A pid from the cgroup or null for all containers
+   * @throws YarnException Could not access cgroups
+   */
+  public CGroupsResourceCalculator(String pid) throws YarnException {
+    this(pid, PROCFS, ResourceHandlerModule.getCGroupsHandler(),
+        SystemClock.getInstance());
+  }
+
+  /**
+   * Create resource calculator for testing.
+   * @param pid A pid from the cgroup or null for all containers
+   * @param procfsDir Path to /proc or a mock /proc directory
+   * @param cGroupsHandler Initialized cgroups handler object
+   * @param clock A clock object
+   * @throws YarnException YarnException Could not access cgroups
+   */
+  @VisibleForTesting
+  CGroupsResourceCalculator(String pid, String procfsDir,
+                            CGroupsHandler cGroupsHandler, Clock clock)
+      throws YarnException {
+    super(pid);
+    this.procfsDir = procfsDir;
+    this.cGroupsHandler = cGroupsHandler;
+    this.pid = pid;
+    this.cpuTimeTracker =
+        new CpuTimeTracker(SysInfoLinux.JIFFY_LENGTH_IN_MILLIS);
+    this.clock = clock;
+    this.jiffyLengthMs = (clock == SystemClock.getInstance()) ?
+      SysInfoLinux.JIFFY_LENGTH_IN_MILLIS : 10;
+    setCGroupFilePaths();
+  }
+
+  @Override
+  public float getCpuUsagePercent() {
+    readTotalProcessJiffies();
+    cpuTimeTracker.updateElapsedJiffies(
+        processTotalJiffies,
+        clock.getTime());
+    return cpuTimeTracker.getCpuTrackerUsagePercent();
+  }
+
+  @Override
+  public long getCumulativeCpuTime() {
```

Comment:
Compatibility. The original CPU time from procfs was based on /proc/<>/stat which are the same counters user+kernel in jiffies.